src/scheduler.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2015 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "scheduler.h"
+
+#include <boost/bind.hpp>
+#include <utility>
+
+CScheduler::CScheduler()
+{
+}
+
+
+boost::thread*
+CScheduler::run()
+{
+    boost::thread* t = new boost::thread(boost::bind(&CScheduler::SchedulerLoop, this));
+    return t;
+}
+
+void
+CScheduler::SchedulerLoop()
+{
+    boost::unique_lock<boost::mutex> lock(newTaskMutex);
+
+    // newTaskMutex is locked throughout this loop EXCEPT
+    // when the thread is waiting or when the user's function
+    // is called.
+    while (1)
+    {
+        if (taskQueue.empty())
+        {
+            // Wait until there is something to do.
+            newTaskScheduled.wait(lock);
```

Comment:
.wait() may wake spuriously, so it's better to turn the surrounding if into a while loop. I believe that means you also won't have to deal with the queue-empty case in the loop below.